### PR TITLE
Fixes #1438 - Make AppDeployIntegrationTest resilient to event reordering

### LINKED ISSUE #1438: AppDeployIntegrationTest is unstable because of event order
Unfortunately, the `AppDeployIntegrationTest` depends on a certain order of events. That does not work.

### PR DESCRIPTION
**Pull request by kolloch**

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 1]
```diff
@@ -44,23 +44,24 @@ class AppDeployIntegrationTest
     deploymentIds.length should be(1)
     val deploymentId = deploymentIds.head
 
-    val apiPostEvent = waitForEvent("api_post_event")
+    log.info("waiting for deployment success")
+    val events: Map[String, Seq[CallbackEvent]] = waitForEvents(
```
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Why do you list events here, that never get accessed in the test afterwards? (e.g. deployment_info)?

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 2]
```diff
@@ -44,23 +44,24 @@ class AppDeployIntegrationTest
...
+      "api_post_event", "group_change_success", "deployment_info",
+      "status_update_event", "status_update_event",
+      "deployment_success")(30.seconds)
```
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: In most other tests, the timeout is handled as optional parameter to enhance readability. We should do that here as well.

File: src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala [hunk 3]
```diff
@@ -52,4 +52,33 @@ trait MarathonCallbackTestSupport extends ExternalMarathonIntegrationTest {
...
+    * Wait for the events of the given kinds (=types).
+    */
+  def waitForEvents(kinds: String*)(maxWait: FiniteDuration = 30.seconds): Map[String, Seq[CallbackEvent]] = {
+
+    val deadline = maxWait.fromNow
+
+    /** Receive the events for the given kinds (duplicates allowed) in any order. */
+    val receivedEventsForKinds: Seq[CallbackEvent] = {
+      var eventsToWaitFor = kinds
+      var receivedEvents: Vector[CallbackEvent] = Vector()
```
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala]: Minor: Vector() -> Vector.empty[CallbackEvent]

**Comment by aquamatthias** [comment]
Integration Tests are running fine.
Added some small comments, but overall LGTM.

File: src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala (hunk 3 above)
Comment (drexin) [inline src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala]: No need to repeat the type param. `Vector.empty` should be enough.

Comment (drexin) [inline src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala]: Or even better remove the type annotation and use `Vector.empty[CallbackEvent]`

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 1 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Even though the test doesn't check that the event info is correct, it checks that the event happens.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 2 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: It is optional in the technical sense as well. you could use empty parenthesis. Of course, you could also move the timeout parameter list before the events but that's not better in my opinion. If it is better in your opinion, I can change that.

Or do you suggest something else?

File: src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala (hunk 3 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala]: Hmmm. Is that important? Is that code, as it is, hard to understand for you? Or what is the problem you see with it?

Comment (drexin) [inline src/test/scala/mesosphere/marathon/integration/setup/MarathonCallbackTestSupport.scala]: Btw. does it have to be a `var`? Can't we use a `Vector.newBuilder` instead?

**Pull request merged**